File: flink-java-core/src/main/java/com/atguigu/flink/day2/KeyedStreamReduceExample.java
package com.atguigu.flink.day2;

import org.apache.flink.api.common.functions.ReduceFunction;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

/**
 * @author ：hmz
 * @date ：Created in 2020/9/11 14:36
 */
public class KeyedStreamReduceExample {

    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        DataStream<SensorReading> stream = env.addSource(new SensorSource());
        stream.filter(i -> i.getId().equals("Sensor_1"))
                .keyBy(i -> i.getId())
                .reduce((r1, r2) -> {
                    if (r1.getTamperature() >= r2.getTamperature()) {
                        return r1;
                    } else {
                        return r2;
                    }
                });

        stream.filter(i -> i.getId().equals("Sensor_1"))
                .keyBy(i -> i.getId())
                .reduce(new MyReduceFunction())
                .print();

        env.execute();
    }

    public static class MyReduceFunction implements ReduceFunction<SensorReading>{

        @Override
        public SensorReading reduce(SensorReading sensorSource, SensorReading t1) throws Exception {
            if(sensorSource.getTamperature() >= t1.getTamperature()){
                return sensorSource;
            }else{
                return t1;
            }
        }
    }

}

File: flink-scala-core/src/main/scala/com/atguigu/flink/day2/KeyedStreamReduceExample.scala
package com.atguigu.flink.day2

import org.apache.flink.streaming.api.scala._

object KeyedStreamReduceExample {

    def main(args: Array[String]): Unit = {

        val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
        env.setParallelism(1)
        val stream: DataStream[SensorReading] = env.addSource(new SensorSource)
        stream.filter(_.id == "Sensor_1")
                        .keyBy(_.id)
                        .reduce((t1,t2)=> new SensorReading(t1.id,t1.timestamp,t1.temperature.max(t2.temperature)))
                        .print()

        env.execute()

    }

}
